### PR TITLE
Build v7 lead scoring dataset with causal trap and canonical validation

### PR DESCRIPTION
**Pull request by shaypal5**
## Summary

- **Purely causal leakage trap**: Removed `boost_leakage_trap()` and all label-conditioned Poisson noise from v6. The v7 trap (`__leakage__touches_post_snapshot_21_90`) counts raw simulated post-snapshot touches — predictive only through shared latent confounders, not label injection.
- **Engine: follow-up ramp mechanism**: `LatentDecayIntensity` gains `followup_boost_after_day`, `followup_boost_factor`, `followup_ramp_days`, and `followup_latent_weights` parameters. After day 20, sales teams increase engagement using *different* latent dimensions (budget_readiness, process_maturity) than pre-snapshot features use (engagement_propensity, account_fit), giving the trap genuinely marginal information.
- **Documentation alignment**: Snapshot day, region values, seniority levels, and all column metadata in RELEASE_v7.md exactly match the generated data (validated programmatically).
- **Honest trap thresholds**: Mean delta ≥ 0.008 (actual: 0.0123), min delta ≥ 0.002 (actual: 0.0048), 10/10 seeds positive. Calibrated for a purely causal trap — no artificial label boost.

### Key metrics

| Metric | Value |
|---|---|
| Baseline LR AUC (seed 42) | 0.625 |
| GBM improvement (5-seed avg) | +0.059 |
| Trap delta (10-seed mean) | +0.0123 |
| Trap delta (10-seed min) | +0.0048 |
| Value-aware uplift K=25 | +38.3% |
| Cohort split AUC drop | 0.066 |
| All mandatory checks | ✅ Pass |

### Files changed

**Engine** (2 files):
- `leadforge/mechanisms/counts.py` — follow-up ramp in `LatentDecayIntensity`
- `leadforge/mechanisms/policies.py` — `_FOLLOWUP_LATENT_WEIGHTS` per motif family

**Pipeline** (1 file):
- `leadforge/pipelines/build_v7.py` — v7 pipeline functions (no `boost_leakage_trap`)

**Scripts** (3 files):
- `scripts/build_v7_snapshot.py` — CLI build script
- `scripts/validate_v7_dataset.py` — full validator with canonical LR pipeline
- `scripts/quick_baseline_eval_v7.py` — multi-model baseline evaluation

**Datasets** (3 files):
- `lead_scoring_intro/lead_scoring_intro_v7.csv` — 1000 × 20 (student)
- `lead_scoring_intro/lead_scoring_intro_v7_instructor.csv` — 1000 × 21 (instructor)
- `lead_scoring_intro/validation_v7_report.json` — full validation output

**Docs** (2 files):
- `lead_scoring_intro/RELEASE_v7.md` — complete release notes
- `lead_scoring_intro/BACKGROUND_v7.md` — student-facing business context

**Tests** (2 files):
- `tests/scripts/test_build_v7_snapshot.py` — 32+ pipeline tests
- `tests/mechanisms/test_mechanisms.py` — 9 new follow-up ramp tests

**CI** (1 file):
- `.github/workflows/ci.yml` — `validate-dataset-v7` job

## Test plan

- [x] All 839 tests pass (`pytest`)
- [x] `ruff check .` clean
- [x] `python scripts/validate_v7_dataset.py` — all mandatory checks pass
- [x] Trap is purely causal: `test_no_label_injection` verifies source code doesn't reference `converted`
- [x] Student/instructor parity: shared columns identical, instructor has exactly 1 trap column
- [ ] CI passes on this PR

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: leadforge/pipelines/build_v7.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
+"""Pipeline functions for building the v7 lead scoring intro CSVs.
```
Comment (shaypal5) [inline leadforge/pipelines/build_v7.py:1]: **Whole-file duplication from v6.**

This file is nearly identical to `build_v6.py` — the only material difference is the removal of `boost_leakage_trap`. Functions like `derive_features`, `softcap_expected_acv`, `assign_acquisition_wave`, `subsample`, `inject_missingness`, and `rename_and_select` are verbatim copies.

The same copy-paste pattern extends to `validate_v7_dataset.py` (~810 lines duplicated from v6) and `quick_baseline_eval_v7.py`.

When a bug is found in `subsample` or `inject_missingness`, it must be fixed in every `build_vN.py` independently. These shared functions belong in `leadforge/pipelines/common.py` with version-specific modules importing from it. The v7 module would then only contain the delta (removal of `boost_leakage_trap`, different constants).

File: leadforge/mechanisms/counts.py [hunk 2]
```diff
@@ -162,6 +180,10 @@ def __init__(
         floor_rate: float = 0.01,
         latent_weights: dict[str, float] | None = None,
         boost: float = 0.8,
+        followup_boost_after_day: int | None = None,
+        followup_boost_factor: float = 1.0,
```
Comment (shaypal5) [inline leadforge/mechanisms/counts.py:184]: **9 constructor parameters — group the follow-up ramp into a dataclass.**

The four `followup_*` parameters are a cohesive unit describing a single concept (ramp behavior). A `FollowupRampConfig` dataclass would:
- Make the call site in `policies.py` readable (one object vs. a wall of kwargs)
- Allow validation of the ramp config in isolation
- Make `None` semantics cleaner: `followup_config: FollowupRampConfig | None = None` vs. 4 independent `None`/default values that must be coordinated

File: leadforge/mechanisms/counts.py [hunk 3]
```diff
@@ -174,19 +196,59 @@ def __init__(
         self._floor = floor_rate
         self._latent_weights: dict[str, float] = dict(latent_weights) if latent_weights else {}
```
Comment (shaypal5) [inline leadforge/mechanisms/counts.py:205]: **No validation on new parameters.**

The constructor validates `base_rate > 0`, `0 < decay_factor <= 1`, and `floor_rate >= 0`, but doesn't validate any of the new follow-up params:
- `followup_ramp_days > 0` (0 gives `max(1, 0) = 1` — works but surprising)
- `followup_boost_factor >= 1.0` (< 1.0 means boost *decreases* over time — may be valid but never tested)
- `followup_boost_after_day >= 0`

This breaks the validation pattern established by existing parameters in the same class.

Also: with `followup_boost_factor=10.0` in `policies.py`, the effective boost at ramp end is `1.2 * 10.0 = 12.0`. Combined with latent multipliers of 2.5+, the rate modulation is `1 + 12 * 2.5 = 31x` base rate. There's no guard or documented rationale for why 10x is reasonable.

File: leadforge/mechanisms/policies.py [hunk 4]
```diff
@@ -216,12 +253,19 @@ def assign_mechanisms(
     touch_intensity: RecencyDecayIntensity | LatentDecayIntensity
     if latent_touch_intensity:
         touch_latent_w = _TOUCH_LATENT_WEIGHTS.get(motif_family, _DEFAULT_TOUCH_LATENT_WEIGHTS)
+        followup_latent_w = _FOLLOWUP_LATENT_WEIGHTS.get(
+            motif_family, _DEFAULT_FOLLOWUP_LATENT_WEIGHTS
+        )
         touch_intensity = LatentDecayIntensity(
             base_rate=touch_rate,
```
Comment (shaypal5) [inline leadforge/mechanisms/policies.py:265]: **Hardcoded magic numbers with no per-motif variation.**

`boost=1.2`, `followup_boost_after_day=20`, `followup_boost_factor=10.0`, `followup_ramp_days=10` are hardcoded inline. Every other mechanism parameter in this file varies by motif family (conversion weights, hazard params, transition weights, touch base rates, latent weights, followup latent weights). But the ramp *dynamics* are identical for all five motifs.

This is inconsistent: a `buying_committee_friction` world presumably has different follow-up timing than a `demo_trial_mediated` world. If this is intentional (same ramp for all motifs), it deserves a comment. If it's an oversight, it needs a per-motif table like everything else.

**Review by shaypal5 — COMMENTED** [review]
Self-review from a skeptical senior-dev perspective. Posting critical remarks to keep us honest.

File: scripts/validate_v7_dataset.py [hunk 5]
```diff
@@ -0,0 +1,809 @@
...
+    return Pipeline(
+        [
+            ("preprocessor", preprocessor),
+            ("classifier", LogisticRegression(max_iter=1000, solver="lbfgs", random_state=42)),
+        ]
+    )
+
+
+def _get_feature_cols(
+    df: pd.DataFrame,
```
Comment (shaypal5) [inline scripts/validate_v7_dataset.py:140]: **Binary features silently dropped from validation pipeline — reported AUCs may be understated.**

`BINARY_FEATURES` (`opportunity_created`, `demo_completed`) are defined at line 66 but `_get_feature_cols()` only returns `cat_cols` and `num_cols`. The binary features are in `FINAL_COLUMNS_STUDENT` but never make it into the sklearn pipeline's feature matrix.

This means the "canonical" LR baseline trains *without* two likely-informative features. The reported AUC (0.625) and all trap deltas are computed on a degraded feature set. This also affects `quick_baseline_eval_v7.py` which has the same pattern.

These should either be added to `NUM_FEATURES` (they're 0/1 ints) or returned as a third group from `_get_feature_cols`.

File: tests/scripts/test_build_v7_snapshot.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+
+# ---------------------------------------------------------------------------
+# Tests — compute_post_snapshot_touches
+# ---------------------------------------------------------------------------
+
+
+class TestComputePostSnapshotTouches:
+    def test_empty_touches(self):
+        snapshot = _make_snapshot(n=10)
```
Comment (shaypal5) [inline tests/scripts/test_build_v7_snapshot.py:354]: **No non-trivial test for `compute_post_snapshot_touches` — the core trap function.**

`test_empty_touches` only tests the trivial case (empty list → all zeros). There is no test that:
- Provides real touch objects and verifies correct day filtering
- Verifies touches on exactly `snapshot_day` are excluded (boundary condition)
- Verifies touches on exactly the horizon boundary are included
- Verifies correct count aggregation per lead

This is the function that computes the leakage trap — the central novelty of v7. It deserves thorough unit tests with non-trivial inputs.

File: tests/scripts/test_build_v7_snapshot.py [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+        result = compute_post_snapshot_touches(snapshot, [], {}, snapshot_day=20)
+        assert (result == 0).all()
+
+    def test_no_label_injection(self):
+        """The trap must NOT use the converted column in any way."""
```
Comment (shaypal5) [inline tests/scripts/test_build_v7_snapshot.py]: **Source-code string matching is fragile — test behavior instead.**

```python
source = inspect.getsource(compute_post_snapshot_touches)
assert ".converted" not in source
```

This misses indirect label access: `df[TARGET]`, a variable holding `"converted"`, column access through a renamed alias, etc. A proper behavioral test would construct two DataFrames with identical features but different `converted` labels, run `compute_post_snapshot_touches` on both, and assert the trap values are identical.

File: leadforge/pipelines/build_v7.py [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+    missing = [c for c in columns if c not in df.columns]
+    if missing:
+        raise ValueError(
+            f"Missing required columns after renaming: {missing}. Available: {sorted(df.columns)}"
+        )
+    return df[columns]
+
+
+def subsample(
+    df: pd.DataFrame,
```
Comment (shaypal5) [inline leadforge/pipelines/build_v7.py:236]: **`subsample` can silently return fewer than `n` rows.**

When there aren't enough negatives (the `n_neg = min(...)` path), the total `n_pos + n_neg` may be less than the requested `n`. The docstring says "Stratified subsample to n rows" but this is not guaranteed. This is the same latent bug inherited from v6 by copy-paste.

The function should either enforce the total count (adjusting both sides) or raise if it can't meet the request.

File: scripts/validate_v7_dataset.py [hunk 9]
```diff
@@ -0,0 +1,809 @@
...
+        lift = prec / base_rate if base_rate > 0 else 0.0
+        metrics[f"precision@{k}"] = prec
+        metrics[f"recall@{k}"] = rec
+        metrics[f"lift@{k}"] = lift
+
+    return errors, metrics
+
+
+def check_tree_improvement(df: pd.DataFrame, label: str) -> tuple[list[str], dict[str, float]]:
+    """Tree model should improve over LR."""
```
Comment (shaypal5) [inline scripts/validate_v7_dataset.py:291]: **ML pipeline defined in 3 places — no single source of truth.**

The "canonical" pipeline appears here (`check_tree_improvement` builds its own GBM pipeline), in `_build_pipeline()` at line 107, and again in `quick_baseline_eval_v7.py` (`_build_preprocessor`). If the canonical pipeline definition changes, it must be updated in all three places.

Also: `check_tree_improvement` has a dead code block:
```python
if improvement < 0.02:
    # Warning, not hard failure
    pass
```
The warning is never emitted — `pass` does nothing. Either remove the dead branch or actually append to `errors`/return a warning.

File: lead_scoring_intro/RELEASE_v7.md [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+The `acquisition_wave` feature enables a distribution-shift lecture:
+- **Random split** approximates IID evaluation (train and test drawn from same distribution)
+- **Cohort split** approximates future generalization: train on waves A+B (earlier leads), test on wave C (latest leads)
+
+| Split | AUC | PR-AUC |
+|---|---|---|
+| Random (70/30, seed 42) | 0.639 | 0.403 |
+| Cohort (A+B → C) | 0.573 | 0.340 |
+| **AUC drop** | **+0.066** | |
+
```
Comment (shaypal5) [inline lead_scoring_intro/RELEASE_v7.md:217]: **"AUC drop" shown as `+0.066` is confusing.**

A "drop" with a positive sign reads as an improvement. This should be `-0.066` (indicating cohort is lower), or the label should be "AUC gap" / "AUC difference (random − cohort)" with an unsigned value.

File: scripts/quick_baseline_eval_v7.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+    print("=" * 60)
+
+    models = {
+        "LR": LogisticRegression(max_iter=1000, solver="lbfgs", random_state=42),
+        "RF": RandomForestClassifier(n_estimators=100, random_state=42),
+        "GBM": GradientBoostingClassifier(n_estimators=100, random_state=42),
+    }
+
+    for name, clf in models.items():
+        aucs = []
```
Comment (shaypal5) [inline scripts/quick_baseline_eval_v7.py:121]: **Same `clf` object reused across seeds.**

The inner loop wraps the same `clf` object in a new `Pipeline` each iteration. sklearn `Pipeline` holds the estimator by reference, so each `fit()` mutates the shared estimator.

The AUC measurements within the loop are correct (fit completes before predict), but `random_state=42` in the classifier means the *tree randomness* is identical across all 5 splits — only the data split varies. This may understate variance.

Consider `clone(clf)` or constructing a fresh estimator per seed.

File: scripts/validate_v7_dataset.py [hunk 12]
```diff
@@ -0,0 +1,809 @@
...
+    "company_size",
+    "company_revenue",
+    "contact_role",
+    "seniority",
+    "lead_source",
+    "acquisition_wave",
+]
+
+BINARY_FEATURES = [
+    "opportunity_created",
```
Comment (shaypal5) [inline scripts/validate_v7_dataset.py:66]: **Feature list duplication — 3 independent definitions.**

`CAT_FEATURES`, `NUM_FEATURES`, `BINARY_FEATURES` are defined independently here, in `quick_baseline_eval_v7.py`, and as `FINAL_COLUMNS_STUDENT` in `build_v7.py`. If a column is added or renamed, all three must be updated.

The validator and eval scripts should import from `leadforge.pipelines.build_v7` (or a shared constants module).
